### DISCUSSION TITLE
Custom styled component - best practices

### DISCUSSION THREAD

**Discussion by huczk** [tamagui/tamagui#635] [Q&A]
Hi, I have a question about custom components with `styled` function. I want to know how to create custom component that works with both - web and native environments. Also, I have problems with typescript and variants.

Let's take snippet below:

```ts
const CustomComponent = (props: { className?: string; test?: boolean; col?: any }) => {
  console.log(props);

  return <Text className={props.className}>hello</Text>;
};

const StyledCustomComponent = styled(CustomComponent, {
  acceptsClassName: true,
  variants: {
    test: {
      true: {
        col: "$gray1",
      },
    },
  } as const,
});
```

Questions:
* Should I add `acceptsClassName` there? I think not, because className is only for web and I want web and native.
* Even with className - values for my variant are not applied, but padding is applied correctly
* Typescript type for props is weird, I must pass my variant content inside props, without it TS will complain about `Types of property 'test' are incompatible.`

**Comment by vjsasinathan** [comment]
Try below approach as mentioned in the docs:
https://tamagui.dev/docs/guides/how-to-build-a-button

Ex:

```
import { Text, styled } from 'tamagui'

const StyledText = Text.styleable((props, ref) => (
  <Text ref={ref} {...props} />
))

export const CustomText = styled(StyledText, {
  name: 'CustomText',
  variants: {
    test: {
      true: {
        col: "$gray1",
      },
    },
  } as const,
})
```
